wgengine/magicsock/magicsock.go

```diff
@@ -635,6 +635,27 @@ func (c *Conn) setEndpoints(endpoints []tailcfg.Endpoint) (changed bool) {
 	return true
 }
 
+// addLatePortMapEndpoint adds a portmapping endpoint to the endpoint set if it could not
+// be created/gotten during determineEndpoints.
+func (c *Conn) addLatePortMapEndpoint(addr netaddr.IPPort) (added bool) {
+	ep := tailcfg.Endpoint{Addr: addr, Type: tailcfg.EndpointPortmapped}
+	c.mu.Lock()
```

Comment:
after acquiring the lock, you need to double check various invariants about the state of the world. Are we even still running, for instance? We might be closed or stopped.

Reply:
I added a check for if it's closed, but are there other invariants to check?